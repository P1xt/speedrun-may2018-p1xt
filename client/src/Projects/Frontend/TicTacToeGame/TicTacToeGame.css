.TicTacToeGame {
  text-align: left;
  margin-top: 40px; }
  .TicTacToeGame > h1 {
    text-transform: none; }
