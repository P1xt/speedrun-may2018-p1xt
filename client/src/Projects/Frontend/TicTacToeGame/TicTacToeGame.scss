.TicTacToeGame {
  text-align: center;
  margin: 40px auto;
  width: 340px;
  &_title {
    margin-top: 1em;
    font-size: 2em;
  }
  &_grid {
    width: 300px;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
  }
  &_cell {
    display: inline-block;
    height: 100px;
    width: 100px;
    border: 1px solid #ffffff;
    text-align: center;
    vertical-align: middle;
    font-size: 4em;
    color: #e27900;
    background: #222222;
  }
  &>h1 {
    text-transform: none
  }

}
