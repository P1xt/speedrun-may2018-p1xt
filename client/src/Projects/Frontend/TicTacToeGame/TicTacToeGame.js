import React from "react";
import { Container, Row, Col } from 'reactstrap';

import "./TicTacToeGame.css";

const TicTacToeGame = () => {
  return (
    <div className="TicTacToeGame">
      <Container>
        <Row>
          <Col>
            <h1>Tic Tac Toe Game</h1>
          </Col>
        </Row>
      </Container>
    </div>
  );
};

export default TicTacToeGame;
