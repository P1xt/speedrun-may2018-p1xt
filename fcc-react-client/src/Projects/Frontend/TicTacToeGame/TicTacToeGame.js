import React, { Component } from "react";
import { Card, CardTitle, CardBody, CardFooter, Button } from "reactstrap";

import "./TicTacToeGame.css";
/**
 * Tic Tac Toe Component
 * @class TicTacToeGame
 * @extends Component
 */
class TicTacToeGame extends Component {
  constructor() {
    super();

    this.state = {
      turn: true,
      values: ["", "", "", "", "", "", "", "", ""],
      text: "New Game"
    };
  }
  /**
   * gameOver
   * checks all possible horizontal, vertical and
   * diagonal wins
   * returns false, "Tie", or 1/input text
   * @memberof TicTacToeGame
   * @param text text to return on gameOver status
   * @param values array of values to test for gameOver
   */
  gameOver = (text = 1, values = this.state.values) => {
    if (
      (values[0] && (values[0] === values[1] && values[0] === values[2])) ||
      (values[3] && (values[3] === values[4] && values[3] === values[5])) ||
      (values[6] && (values[6] === values[7] && values[6] === values[8])) ||
      (values[0] && (values[0] === values[3] && values[0] === values[6])) ||
      (values[1] && (values[1] === values[4] && values[1] === values[7])) ||
      (values[2] && (values[2] === values[5] && values[2] === values[8])) ||
      (values[0] && (values[0] === values[4] && values[0] === values[8])) ||
      (values[2] && (values[2] === values[4] && values[2] === values[6]))
    ) {
      return text;
    } else if (values.reduce((a, b) => a && b, true)) {
      return "Tie!";
    }
    return false;
  };

  /**
   * processTurn
   * handles player click, registers selection
   * to state, and checks if game over
   * @memberof TicTacToeGame
   * @param position position player selected
   */
  processTurn = position => {
    if (this.state.turn === true) {
      const values = this.state.values;
      if (!this.state.values[position]) {
        values[position] = "x";
        this.setState({
          values: values,
          turn: false
        });
      }

      if (this.gameOver())
        this.setState({ turn: false, text: this.gameOver("You Win") });
      else this.processAITurn();
    }
  };

  /**
   * primePosition
   * check possible move for AI, return if it
   * will put the AI in a position to win/block
   * @memberof TicTacToeGame
   */
  primePosition = values => {
    return this.gameOver("WIN", values) === "WIN";
  };

  /**
   * processAITurn
   * select and make optinal move
   * on AI turn
   * @memberof TicTacToeGame
   */
  processAITurn = () => {
    let values = [...this.state.values];
    let moved = false;
    for (let i = 0; i < 9; i += 1) {
      // check each position
      if (!values[i]) {
        // if AI can win, do so
        values[i] = "o";
        if (this.primePosition(values)) {
          moved = true;
          this.setState({
            values: [...values],
            text: "AI Wins",
            turn: false
          });
        } else {
          // if AI can block player win, do so
          values[i] = "x";
          if (this.primePosition(values)) {
            values[i] = "o";
            moved = true;
            if (this.gameOver(values)) {
              this.setState({
                text: this.gameOver("AI Wins"),
                turn: false
              });
            } else {
              this.setState({
                values: [...values],
                turn: true
              });
            }
          }
        }
        values[i] = "";
      }
    }

    // if AI couldn't win or block win
    // take first position available
    for (let i = 0; i < 9 && moved === false; i += 1) {
      if (!values[i]) {
        values[i] = "o";
        this.setState({
          values: values,
          turn: true
        });
        moved = true;
      }
    }

    return;
  };

  /**
   * reset
   * reset game
   * @memberof TicTacToeGame
   */
  reset = () =>
    this.setState({
      values: ["", "", "", "", "", "", "", "", ""],
      text: "New Game",
      turn: true
    });

  /**
   * render
   * render Tic Tac Toe Component
   * @return
   * @memberof TicTacToeGame
   */
  render() {
    return (
      <div className="TicTacToeGame">
        <Card>
          <CardTitle className="TicTacToeGame_title">Tic Tac Toe</CardTitle>
          <CardBody>
            <div className="TicTacToeGame_grid">
              {[...Array(9)].map((x, i) => (
                <div
                  className="TicTacToeGame_cell"
                  key={i}
                  onClick={() => this.processTurn(i)}
                >
                  {this.state.values[i]}
                </div>
              ))}
            </div>
          </CardBody>
          <CardFooter>
            <Button color="info" onClick={this.reset}>
              {this.state.text}
            </Button>
          </CardFooter>
        </Card>
      </div>
    );
  }
}

export default TicTacToeGame;
