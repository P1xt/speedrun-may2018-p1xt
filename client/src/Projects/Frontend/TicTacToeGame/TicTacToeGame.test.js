import React from "react";
import renderer from 'react-test-renderer';
import { shallow } from "enzyme";

import TicTacToeGame from "./TicTacToeGame";

describe("<TicTacToeGame />", () => {
  it("renders without crashing", () => {
    shallow(<TicTacToeGame />);
  });
  it("renders correctly", () => {
    const tree = renderer.create(<TicTacToeGame />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
