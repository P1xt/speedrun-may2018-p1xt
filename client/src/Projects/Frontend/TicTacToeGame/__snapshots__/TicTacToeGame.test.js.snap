// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicTacToeGame /> renders correctly 1`] = `
<div
  className="TicTacToeGame"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <h1>
          Tic Tac Toe Game
        </h1>
      </div>
    </div>
  </div>
</div>
`;
